src/src/context/TodosContext.tsx

```diff
@@ -0,0 +1,207 @@
+import React, { createContext, useEffect, useState } from 'react';
+import { Todo } from '../types/Todo';
+import { Statuses } from '../types/Statuses';
+import { getTodos } from '../api/todos';
+import { USER_ID } from '../constants/USER_ID';
+import { Errors } from '../types/Errors';
+import * as TodoService from '../api/todos';
+
+type PropsTodosContext = {
+  todos: Todo[];
+  setTodos: (todos: Todo[]) => void;
+  filter: Statuses;
+  setFilter: (value: Statuses) => void;
+  errorMessage: Errors,
+  setErrorMessage: (value: Errors) => void,
+  tempTodo: Todo | null,
+  setTempTodo: (tempTodo: Todo | null) => void,
+  title: string,
+  setTitle: (title: string) => void,
+  isLoading: boolean,
+  setIsLoading: (value: boolean) => void,
+  addTodo: (todo: Todo) => void,
+  deleteTodo: (todoId: number) => void,
+  deleteCompletedTodos: () => void;
+  upgradeTodo: (todo: Todo) => void,
+  handleToggleAll: () => void,
+};
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const TodosContext = createContext<PropsTodosContext>({
+  todos: [],
+  setTodos: () => { },
+  filter: Statuses.All,
+  setFilter: () => {},
+  errorMessage: Errors.NoErrors,
+  setErrorMessage: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  title: '',
+  setTitle: () => {},
+  isLoading: false,
+  setIsLoading: () => {},
+  addTodo: () => {},
+  deleteTodo: () => {},
+  deleteCompletedTodos: () => {},
+  upgradeTodo: () => {},
+  handleToggleAll: () => {},
+});
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filter, setFilter] = useState<Statuses>(Statuses.All);
+  const [title, setTitle] = useState<string>('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [errorMessage, setErrorMessage] = useState<Errors>(Errors.NoErrors);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
```

Comment:
loader should be visible on eact todo which in some process on server. if you editing or deleting multiply todos - loader should be visible on each of them.
for it suggest instead boolean state isLoading make some state which will be manage all todos which in process. it can be array of ids such todos. 
please compare this moment with [working example](https://mate-academy.github.io/react_todo-app-with-api/)

Reply:
It's a brilliant idea.